plugin/pkg/controllerregistration/resources/admission.go

```diff
@@ -0,0 +1,133 @@
+func NewFactory(config io.Reader) (admission.Interface, error) {
+	return New()
+}
+
+// Resources contains an admission handler and listers.
+type Resources struct {
+	*admission.Handler
+	coreClient coreclientset.Interface
+	readyFunc  admission.ReadyFunc
+}
+
+var (
+	_ = admissioninitializer.WantsInternalCoreClientset(&Resources{})
+
+	readyFuncs = []admission.ReadyFunc{}
+)
+
+// New creates a new Resources admission plugin.
+func New() (*Resources, error) {
+	return &Resources{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+	}, nil
+}
+
+// AssignReadyFunc assigns the ready function to the admission handler.
+func (r *Resources) AssignReadyFunc(f admission.ReadyFunc) {
+	r.readyFunc = f
+	r.SetReadyFunc(f)
+}
+
+// SetInternalCoreClientset gets the clientset from the Kubernetes client.
+func (r *Resources) SetInternalCoreClientset(c coreclientset.Interface) {
+	r.coreClient = c
+}
+
+// ValidateInitialization checks whether the plugin was correctly initialized.
+func (r *Resources) ValidateInitialization() error {
+	return nil
+}
+
+// Validate makes admissions decisions based on the resources specified in a ControllerRegistration object.
+// It does reject the request if there is any other existing ControllerRegistration object in the system that
+// specifies the same resource kind/type combination like the incoming object.
+func (r *Resources) Validate(a admission.Attributes) error {
+	// Wait until the caches have been synced
+	if r.readyFunc == nil {
+		r.AssignReadyFunc(func() bool {
+			for _, readyFunc := range readyFuncs {
+				if !readyFunc() {
+					return false
+				}
+			}
+			return true
+		})
+	}
+	if !r.WaitForReady() {
+		return admission.NewForbidden(a, errors.New("not yet ready to handle request"))
+	}
+
+	// Ignore all kinds other than Shoot or Project.
+	// TODO: in future the Kinds should be configurable
+	// https://v1-9.docs.kubernetes.io/docs/admin/admission-controllers/#imagepolicywebhook
+	if a.GetKind().GroupKind() != core.Kind("ControllerRegistration") {
+		return nil
+	}
+	controllerRegistration, ok := a.GetObject().(*core.ControllerRegistration)
+	if !ok {
+		return apierrors.NewBadRequest("could not convert resource into ControllerRegistration object")
+	}
+
+	// Live lookup to prevent missing any data
+	controllerRegistrationList, err := r.coreClient.Core().ControllerRegistrations().List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	existingResources := map[string]string{}
+	for _, obj := range controllerRegistrationList.Items {
+		for _, resource := range obj.Spec.Resources {
+			existingResources[resource.Kind] = resource.Type
+		}
+	}
+
+	for _, resource := range controllerRegistration.Spec.Resources {
+		if t, ok := existingResources[resource.Kind]; ok && t == resource.Type {
+			return admission.NewForbidden(a, fmt.Errorf("another ControllerRegistration resource already exists that supports resource %s/%s", resource.Kind, resource.Type))
```

Comment:
Why this is needed? Normally, multiple controllers can operate on the same resource.